### DISCUSSION TITLE
Rails and ActiveSupport Notes

### DISCUSSION THREAD

**Discussion by ohler55** [ohler55/oj#739] [General]
Please place notes about dealing with Rails and ActiveSupport in this discussion topic.

**Comment by RamilGN** [comment]
What is the main difference between `Oj.dump` and `Oj::Rails.encode?`

**Reply by HeyNonster** [reply]
`Oj.dump`, using the default settings, will not return the same format that Rails' `ActiveSupport` will.

```
ActiveSupport::JSON.encode(Time.now.utc)                      => "2022-01-20T09:32:27.708Z"
Oj.dump(Time.now.utc)                                         => {"^t":1642671147.709}
```

However, `Oj::Rails.encode` will mimic the format of `ActiveSupport`:

```
ActiveSupport::JSON.encode(Time.now.utc)                      => "2022-01-20T09:32:27.708Z"
Oj::Rails.encode(Time.now.utc)                                => "2022-01-20T09:32:27.708Z"
```

But if you set `Oj` with `mode: :rails` then `Oj.dump` _will also_ mimic `ActiveSupport`.

```
ActiveSupport::JSON.encode(Time.now.utc)                      => "2022-01-20T09:36:15.727Z"
Oj.dump(Time.now.utc)                                         => "2022-01-20T09:36:15.727Z"
```

Basically, if you don't want to set `mode: :rails` but still want `Oj` to return the same format as `ActiveSupport` then you can use `Oj::Rails.encode`. I'm pretty sure that's how it works, at least.

**Reply by ohler55** [reply]
The interaction between Rails and the JSON gem can be somewhat unpredictable. Oj does it's best to reproduce that. As pointed out `Oj.dump` is different than the ActiveSupport calls and different than the JSON gem although with the options set correctly it can give the same results. The options are the key.

**Comment by HeyNonster** [comment]
Hey, everyone, this write-up mostly relates to a caveat when using `Oj version <= 3.10.0` with `Rails/ActiveSupport` and `Oj.optimize_rails` but may also provide insight into what `Oj` is doing under the hood if that's of interest. 

**TL/DR: In`Oj version <= 3.10.0` passing a time object to `JSON.dump` and passing a time object nested inside an array or hash to `JSON.dump` will not return the same time format.  This is because of all the monkey patching that ActiveSupport does and was fixed in `Oj 3.10.1`**

Extended details below:

My understanding is that with `Oj.optimize_rails` Oj will replace `JSON.dump` && `ActiveSupport::JSON.encode` with faster implementations written in C. Without `Oj.optimize_rails` Oj will provide fast `Oj.dump` methods but otherwise leave everything else alone.

Here's a demonstration with these Oj settings: `Oj.default_options = { mode: :rails, trace: true }` and with `ActiveSupport`

```ruby
# Oj version 3.10.0

#JSON.dump
JSON.dump(Time.now.utc)                                => "2022-01-11 12:17:36 UTC"

# ActiveSupport::JSON.encode
ActiveSupport::JSON.encode(Time.now.utc)               => "2022-01-11T12:17:36.614Z"

# Oj.dump
#0:      rails.c:1468:Oj:}: dump Time
#0:      rails.c:517:Oj:>:   as_json Time
#0:      rails.c:527:Oj:<:   as_json Time
#0:      rails.c:1468:Oj:}: dump String
#0:      rails.c:1479:Oj:{: dump String
#0:      rails.c:1479:Oj:{: dump Time
Oj.dump(Time.now.utc)                                  => "2022-01-11T12:17:36.614Z"
```

Notice above the `Oj.dump(Time.now.utc)` line. The `trace: true` Oj option gives you a window into what Oj is doing and you can see here that `Oj.dump` utilizes the fast C code. Notably, the file is `rails.c`. This is because we set `mode: :rails` in the `default_options` which tells Oj to render the JSON in the same format that ActiveSupport would.

Now, if we run the same code but add `Oj.optimize_rails`:

```ruby
# Oj version 3.10.0

#JSON.dump
#0:dump_compat.c:940:Oj:}: dump Time
#0:dump_compat.c:965:Oj:{: dump Time
JSON.dump(Time.now.utc)                                => "2022-01-11 12:27:17 UTC"

# ActiveSupport::JSON.encode
#0:      rails.c:1468:Oj:}: dump Time
#0:      rails.c:1479:Oj:{: dump Time
ActiveSupport::JSON.encode(Time.now.utc)               => "2022-01-11T12:27:17.051Z"

# Oj.dump
#0:      rails.c:1468:Oj:}: dump Time
#0:      rails.c:1479:Oj:{: dump Time
Oj.dump(Time.now.utc)                                  => "2022-01-11T12:27:17.051Z"
```

Both `JSON.dump` and `ActiveSupport::JSON.encode` now run through faster C implementations. `JSON.dump`,
 however runs through the file `dump_compat.c` instead of `rails.c`. This is intended so that `JSON.dump` with Oj and `JSON.dump` _without_ Oj return the same time format, the Oj implementation is just faster.
 
The bug in `Oj < 3.10.1` occured when you passed a Hash or Array with time objects to `JSON.dump`
 
```ruby
# Oj version 3.10.0

# JSON.dump
#0:dump_compat.c:940:Oj:}: dump Hash
#0:dump_compat.c:940:Oj:}:   dump Time
#0:dump_compat.c:123:Oj:>: to_json Time
#0:      rails.c:1468:Oj:}: dump Time
#0:      rails.c:1479:Oj:{: dump Time
#0:dump_compat.c:131:Oj:<: to_json Time
#0:dump_compat.c:965:Oj:{:   dump Time
#0:dump_compat.c:965:Oj:{: dump Hash
JSON.dump({created_at: Time.now.utc})                  => {"created_at":"2022-01-11T12:27:16.966Z"}
```

As you can see in the trace, `JSON.dump({created_at: Time.now.utc})` was running the code through both `dump_compat.c` _and_ `rails.c` and returning the Time in the ActiveSupport format when it was supposed to return the same format that the JSON standard library gem would.

This was fixed in `Oj 3.10.1` as you can see in the below trace:

```ruby
# Oj version 3.10.1

# JSON.dump
#0:dump_compat.c:941:Oj:}: dump Hash
#0:dump_compat.c:941:Oj:}:   dump Time
#0:dump_compat.c:123:Oj:>: to_json Time
#0:dump_compat.c:131:Oj:<: to_json Time
#0:dump_compat.c:966:Oj:{:   dump Time
#0:dump_compat.c:966:Oj:{: dump Hash
JSON.dump({created_at: Time.now.utc})                  => {"created_at":"2022-01-11 12:34:36 UTC"}
```

`JSON.dump({created_at: Time.now.utc})` now exclusively uses the `dump_compat.c` file which was the intended behavior. Unfortunately, we had some code that relied on the time format produced by the buggy version of JSON.dump and that broke some things when we upgraded Oj.

Because Oj has to mimic the JSON standard library gem time format and, separately, mimic the ActiveSupport time format (both of which Oj has no control over, it just needs to adapt) we might write a test that catches whenever one the formats change.
